metric_learn/mmc.py

```diff
@@ -344,10 +340,11 @@ def _D_constraint(self, X, c, d, w):
     diff = X[c] - X[d]
     diff_sq = diff * diff
     dist = np.sqrt(diff_sq.dot(w))
-    sum_deri1 = np.sum(diff_sq / (2 * np.maximum(dist, 1e-6))[:,None], axis = 0)
-    sum_deri2 = np.sum(
-      np.einsum('ij,ik->ijk', diff_sq, diff_sq) / (-4 * np.maximum(1e-6, dist**3)[:,None,None]),
-      axis = 0
+    sum_deri1 = np.einsum('ij,i', diff_sq, 0.5 / np.maximum(dist, 1e-6))
+    sum_deri2 = np.einsum(
+      'ijk,i',
+      np.einsum('ij,ik->ijk', diff_sq, diff_sq),
+      -0.25 / np.maximum(1e-6, dist**3)
```

Comment:
Neither do I think so. However, this version is about 4 times faster, since it does not require broadcasting `dist` from 1-D to 3-D. So, I will just pull out the inner `einsum` to a local variable.